content/post/2025-03-17-htsget-rs-in-depth/2025-03-17-htsget-rs-in-depth.md

```diff
@@ -0,0 +1,209 @@
+
+## Querying the header
+
+As a simple example, query the header of the file example file by passing `class=Header`:
+
+```sh
+htsget "https://htsget.ga4gh-demo.org/reads/htsnexus_test_NA12878?class=header" > out.bam
+```
+
+Internally, this yields a JSON with a URL that can be fetched along with a "Range" header:
+
+```json
+{
+  "htsget": {
+    "format": "BAM",
+    "urls": [
+      {
+        "url": "...",
+        "headers": {
+          "Range": "bytes=0-4667"
+        },
+        "class": "header"
+      }
+    ]
+  }
+}
+```
+
+The client takes care of fetching the URLs and concatenating bytes.
+
+A strength of the htsget protocol is that the output represents a small part of the full file, allowing the user to
+query specific regions of a file without needing to obtain the entire file.
+
+In this case, the output represents the BAM header of the file:
+
+```sh
+samtools view -H out
+```
+
+Note that "..." inside the JSON example responses represents some data or a URL. This will be different when executing
+the query.
+
+## Querying reference names with start and end ranges
+
+A more interesting query would involve selecting a specific region, for example chr11. This can be accomplished by
+using the `referenceName` parameter. Viewing the output will show data for that specific region:
+
+```sh
+htsget "https://htsget.ga4gh-demo.org/reads/htsnexus_test_NA12878?referenceName=11" | samtools view
+```
+
+Similarly, the query can be refined further by specifying specific start and end ranges, so that only those regions
+are returned:
+
+```sh
+htsget "https://htsget.ga4gh-demo.org/reads/htsnexus_test_NA12878?referenceName=11&start=500000&end=5001000" | samtools view
+```
+
+Internally, the output from htsget-rs will contain multiple URL tickets that represent the specific data queried:
+
+```json
+{
+  "htsget": {
+    "format": "BAM",
+    "urls": [
+      {
+        "url": "...",
+        "headers": {
+          "Range": "bytes=0-273085"
+        }
+      },
+      {
+        "url": "...",
+        "headers": {
+          "Range": "bytes=499249-574358"
+        }
+      },
+      {
+        "url": "...",
+        "headers": {
+          "Range": "bytes=627987-647345"
+        }
+      },
+      {
+        "url": "...",
+        "headers": {
+          "Range": "bytes=824361-842100"
+        }
+      },
+      {
+        "url": "...",
+        "headers": {
+          "Range": "bytes=977196-996014"
+        }
+      },
+      {
+        "url": "...",
+        "headers": {
+          "Range": "bytes=2596771-2596798"
+        }
+      }
+    ]
+  }
+}
+```
+
+## Querying Crypt4GH files
+
+Moving on to a more complex example, we will now incorporate querying [Crypt4GH][c4gh] encrypted files from htsget-rs.
+To decrypt Crypt4GH files, install the Crypt4GH [CLI][c4gh-cli] and get the [keys] from the htsget-rs repository.
+
+Then, query like before, except add the `encryptionScheme` parameter:
```

Comment:
Remind that this is an experimental feature, reference spec PR: https://github.com/samtools/hts-specs/pull/808